Refresh.PresenceServer/Server/PresenceServer.cs

```diff
@@ -0,0 +1,303 @@
+#if NET9_0_OR_GREATER
+#error Please remove the ToArray call here!
+#endif
+            
+            // Wait for all receive tasks to finish
+            Task.WaitAll(receiveTasks.Concat(sendTasks).ToArray());
+
+            // Clear out the receive task references
+            foreach (GameClient client in this._clients) 
+                client.ReceiveTask = null;
+
+            // Sleep for 1 second
+            await Task.Delay(1000);
+        }
+    }
+
+    private int EncryptSlotId(int slotId)
+    {
+        Span<byte> span = stackalloc byte[sizeof(int)];
+
+        BinaryPrimitives.WriteInt32BigEndian(span, slotId);
+
+        ResourceHelper.XxteaEncrypt(span, this._key);
+        
+        return BinaryPrimitives.ReadInt32BigEndian(span);
+    }
+    
+    private async Task ReceiveTask(object? state)
+    {
+        GameClient gameClient = (GameClient)state!;
+        
+        try
+        {
+            if (gameClient.TcpClient.Available == 0)
+                return;
+
+#if NET9_0_OR_GREATER
+#error Please clean this mess to use Span<T>!!!
+#endif
+
+            int readAmount = await gameClient.TcpClient.Client.ReceiveAsync(gameClient.ReceiveBuffer);
+            if (readAmount == 0)
+                return;
+
+            byte[] read = gameClient.ReceiveBuffer[..readAmount];
+
+            switch (read[0])
+            {
+                // login packet
+                case 0x4c when read[1] == 0x0d && read[2] == 0x0a:
+                {
+                    // Decrypt the body of the packet
+                    ResourceHelper.XxteaDecrypt(read.AsSpan()[3..][..128], this._key);
+
+                    // Convert the auth token back into a string
+                    string authToken = Encoding.UTF8.GetString(read.AsSpan()[3..][..128]["MM_AUTH=".Length..]).TrimEnd('\0');
+
+                    this._logger.LogInfo(PresenceCategory.Authentication, "{0} logged in", gameClient.IpAddress,
+                        authToken);
+
+                    // Set the user's auth token
+                    gameClient.AuthToken = authToken;
+
+                    this._authenticatedClients[authToken] = gameClient;
+
+                    _ = Task.Run(async () =>
+                    {
+                        bool success = await this._apiClient.InformConnection(authToken);
+                        
+                        if(!success)
+                            this._toRemove.Enqueue(gameClient);
+                    });
+                    
+                    break;
+                }
+                // keepalive packet
+                case 0x0d when read[1] == 0x0a && read.Length == 2:
+                    this._logger.LogDebug(PresenceCategory.Connections, "Keepalive from {0}", gameClient.IpAddress);
+
+                    break;
+                default:
+                    this._logger.LogWarning(PresenceCategory.Connections,
+                        "Unknown packet from {0}, treating as basic keepalive", gameClient.IpAddress);
```

Comment:
I feel like we should log the packet in debug at least? Probably fine though

Reply:
Given the packets logged are the only ones the client is ever capable of sending, other packet data is an indication that something has gone quite wrong (eg. HTTP requests are somehow making their way through to it), so I'd prefer to keep it a warning, we can move to debug only if it starts happening frequently though